src/main/java/edu/stanford/protege/metaproject/impl/ConfigurationBuilder.java

```diff
@@ -0,0 +1,843 @@
+        }
+        return this;
+    }
+
+    /**
+     * Change the display name of the given user
+     *
+     * @param userId    User identifier
+     * @param userName  New name
+     * @return ServerConfigurationBuilder
+     */
+    public ConfigurationBuilder setUserName(UserId userId, Name userName) {
+        checkNotNull(userId);
+        checkNotNull(userName);
+        getUser(userId).ifPresent(user -> setUser(userId, factory.getUser(userId, userName, user.getEmailAddress())));
+        return this;
+    }
+
+    /**
+     * Change the email address of a user
+     *
+     * @param userId  User identifier
+     * @param emailAddress New email address
+     * @return ServerConfigurationBuilder
+     */
+    public ConfigurationBuilder setUserEmailAddress(UserId userId, EmailAddress emailAddress) {
+        checkNotNull(userId);
+        checkNotNull(emailAddress);
+        getUser(userId).ifPresent(user -> setUser(userId, factory.getUser(userId, user.getName(), emailAddress)));
+        return this;
+    }
+
+    private Optional<User> getUser(UserId userId) {
+        for(User u : users) {
+            if(u.getId().equals(userId)) {
+                return Optional.of(u);
+            }
+        }
+        return Optional.empty();
+    }
+
+
+    /* projects */
+
+    /**
+     * Add the given project to the configuration
+     *
+     * @param project   New project
+     * @throws IdAlreadyInUseException  Project identifier already in use
+     * @return ServerConfigurationBuilder
+     */
+    public ConfigurationBuilder addProject(Project project) throws IdAlreadyInUseException {
+        checkNotNull(project);
+        if (projects.contains(project)) {
+            throw new IdAlreadyInUseException("The specified project identifier is already used by another project");
+        }
+        projects.add(project);
+        return this;
+    }
+
+    /**
+     * Remove the given project from the configuration
+     *
+     * @param project   Project
+     * @return ServerConfigurationBuilder
+     */
+    public ConfigurationBuilder removeProject(Project project) {
+        checkNotNull(project);
+        projects.remove(project);
+        return this;
+    }
+
+    /**
+     * Modify the project with the specified project identifier with the given project
+     *
+     * @param projectId Project identifier
+     * @param project   Project
+     * @return ServerConfigurationBuilder
+     */
+    public ConfigurationBuilder setProject(ProjectId projectId, Project project) {
+        checkNotNull(projectId);
+        checkNotNull(project);
+        getProject(projectId).ifPresent(this::removeProject);
+        try {
+            addProject(project);
+        } catch (IdAlreadyInUseException e) {
+            logger.debug("Attempted to replace a project but the original project instance was not properly deleted beforehand");
```

Comment:
hmm, confusing. If the delete fails (id doesn't exist), then the add will succeed, non? If the add can only fail if the delete fails, and the delete fails when the id does not exist, then the exception doesn't seem right.

But still isn't it the case that to the caller the set does not actually happen?